### PR TITLE
miscellaneous type system improvements

### PR DESCRIPTION
**Pull request by lcnr**
see review comments for rationale

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_infer/src/infer/relate/generalize.rs [hunk 1]
```diff
@@ -287,6 +299,49 @@ impl<'tcx> Generalizer<'_, 'tcx> {
...
+                            span: self.span,
+                        },
+                        self.for_universe,
+                    ))
+                }
+            }
+        };
+        self.in_alias = is_nested_alias;
+        result
+    }
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/relate/generalize.rs:344]: extracts the alias code into a separate method. generally useful as it's quite involved and also part of #119106

File: compiler/rustc_trait_selection/src/solve/normalize.rs [hunk 2]
```diff
@@ -85,25 +85,16 @@ impl<'tcx> NormalizationFolder<'_, 'tcx> {
             ),
         );
 
-        // Do not emit an error if normalization is known to fail but instead
-        // keep the projection unnormalized. This is the case for projections
-        // with a `T: Trait` where-clause and opaque types outside of the defining
-        // scope.
-        let result = if infcx.predicate_may_hold(&obligation) {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/normalize.rs:92]: by using `AliasRelate` rigid projections don't result in errors anymore

File: compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs [hunk 3]
```diff
@@ -18,8 +18,9 @@ use rustc_middle::ty::{self, Ty, TyCtxt};
 use rustc_middle::ty::{ToPredicate, TypeVisitableExt};
 use rustc_span::{sym, ErrorGuaranteed, DUMMY_SP};
 
+mod anon_const;
 mod inherent;
-mod opaques;
+mod opaque_types;
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs:23]: match the name for all other opaque types files in rustc

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**